### PR TITLE
:tada: Add S3 and GCS stagings for Azure Blob Storage destination

### PR DESCRIPTION
**Pull request by andriikorotkov**
## What
Add S3 and GCS stagings for Azure Blob Storage destination

## Recommended reading order
1. `x.java`
2. `y.python`

## 🚨 User Impact 🚨
No impact

## Pre-merge Checklist
Expand the relevant checklist and delete the others. 

<details><summary> <strong> New Connector </strong></summary>
<p>

#### Community member or Airbyter
   
- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] `docs/SUMMARY.md`
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the connector is published, connector added to connector index as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
   
</p>
</details>

<details><summary> <strong> Updating a connector </strong></summary>
<p>
   
#### Community member or Airbyter
   
- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the new connector version is published, connector version bumped in the seed directory as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</p>
</details>

<details><summary> <strong> Connector Generator </strong> </summary>
<p>
   
- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed.
</p>
</details>

---
This change is [Reviewable](https://reviewable.io/reviews/airbytehq/airbyte/9336)

### CODE REVIEW

File: airbyte-integrations/connectors/destination-azure-blob-storage/src/main/java/io/airbyte/integrations/destination/azure_blob_storage/file/uploader/AbstractS3AzureUploader.java [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+import io.airbyte.protocol.models.DestinationSyncMode;
+import java.time.LocalDateTime;
+import java.time.ZoneId;
+import java.util.Date;
+import java.util.List;
+import java.util.function.Consumer;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public abstract class AbstractS3AzureUploader<T extends S3Writer> extends AbstractAzureUploader<S3Writer> {
```
Comment (alexandr-shegeda) [inline airbyte-integrations/connectors/destination-azure-blob-storage/src/main/java/io/airbyte/integrations/destination/azure_blob_storage/file/uploader/AbstractS3AzureUploader.java:24]: why do we need to have both AbstractS3AzureUploader and AbstractGcsAzureUploader? they look like duplicates with a small difference in the AmazonS3 client initialization

File: airbyte-integrations/connectors/destination-azure-blob-storage/src/main/java/io/airbyte/integrations/destination/azure_blob_storage/AzureBlobStorageDestination.java [hunk 2]
```diff
@@ -32,6 +50,28 @@ public AirbyteConnectionStatus check(final JsonNode config) {
...
       final AzureBlobStorageConnectionChecker client = new AzureBlobStorageConnectionChecker(
           AzureBlobStorageDestinationConfig.getAzureBlobStorageConfig(config));
+      final UploadingMethod uploadingMethod = AzureUtils.getLoadingMethod(config);
+      final UploaderType uploaderType = AzureUtils.getUploaderType(config);
+
+      // GCS upload time re-uses destination-GCS for check and other uploading (CSV format writer)
+      if (UploadingMethod.GCS.equals(uploadingMethod)) {
+        final GcsDestination gcsDestination = new GcsDestination();
+        final JsonNode gcsJsonNodeConfig = AzureUtils.getStagingJsonConfig(uploadingMethod, uploaderType, config);
+        final AirbyteConnectionStatus airbyteConnectionStatus = gcsDestination.check(gcsJsonNodeConfig);
```
[the remark this answers is not in the document]
Comment (andriikorotkov) [reply airbyte-integrations/connectors/destination-azure-blob-storage/src/main/java/io/airbyte/integrations/destination/azure_blob_storage/AzureBlobStorageDestination.java]: updated it

File: airbyte-integrations/connectors/destination-azure-blob-storage/src/main/java/io/airbyte/integrations/destination/azure_blob_storage/file/uploader/AbstractS3AzureUploader.java (hunk 1 above)
Comment (andriikorotkov) [reply airbyte-integrations/connectors/destination-azure-blob-storage/src/main/java/io/airbyte/integrations/destination/azure_blob_storage/file/uploader/AbstractS3AzureUploader.java:14]: updated it

**Review by DoNotPanicUA — APPROVED** [review]

**Comment by tuliren** [comment]
I am not sure why we want to add S3 and GCS loading method for Azure blob storage. The purpose of the Azure blob storage destination is to store objects directly on Azure. Adding the S3 or GCS loading method unnecessarily copy the data first to S3 or GCS, and then to Azure.

This PR is associated with https://github.com/airbytehq/airbyte/issues/6063. However, it does not address the issue. The author of that issue wants us to put a timestamp in the object name (so that users can incrementally load the objects to their data warehouse). Routing the data first to S3 or GCS and then to Azure should not be the ask of that issue.

How was this project scoped in the first place? There was probably some miscommunication there.

**Comment by andriikorotkov** [comment]
@tuliren, @alexandr-shegeda Yes, there seems to be a miscommunication. Please tell me your vision for this issue. Are you suggesting adding a timestamp to the name of the blob (stream airbyte name) and creating a new blob each time with a new timestamp? If so, will it be relevant for all of our users who use Azure Blob Storage destination? Or is it necessary to add some flag to the spec that will switch the logic of work? If we draw an analogy with S3 and GCS, the files of each stream are in a separate folder with the name of the stream. Should I create Azure container for each stream or leave the old logic and write everything in one container? If you leave the old logic, then this will greatly complicate the deletion of files when using overwrite synch mode.

**Comment by tuliren** [comment]
Replied in https://github.com/airbytehq/airbyte/issues/6063.

**Pull request closed without merging**